src/dex/curve-v1-factory/curve-v1-factory.ts

```diff
@@ -1175,31 +1206,91 @@ export class CurveV1Factory
     data: CurveV1FactoryData,
     side: SwapSide,
   ): DexExchangeParam {
-    if (!this.buySideSupported && side === SwapSide.BUY)
-      throw new Error(`Buy not supported`);
+    if (data.path.length === 1) {
+      const { exchange, i, j, underlyingSwap } = data.path[0];
 
-    const { exchange, i, j, underlyingSwap } = data;
+      const minAmountToReceive =
+        side === SwapSide.SELL ? MIN_AMOUNT_TO_RECEIVE : destAmount;
+      const defaultArgs = [i, j, srcAmount, minAmountToReceive];
 
-    const minAmountToReceive =
-      side === SwapSide.SELL ? MIN_AMOUNT_TO_RECEIVE : destAmount;
-    const defaultArgs = [i, j, srcAmount, minAmountToReceive];
+      const swapMethod = underlyingSwap
+        ? CurveSwapFunctions.exchange_underlying
+        : CurveSwapFunctions.exchange;
 
-    const swapMethod = underlyingSwap
-      ? CurveSwapFunctions.exchange_underlying
-      : CurveSwapFunctions.exchange;
-    const exchangeData = this.ifaces.exchangeRouter.encodeFunctionData(
-      swapMethod,
-      defaultArgs,
-    );
+      const exchangeData = this.ifaces.exchangeRouter.encodeFunctionData(
+        swapMethod,
+        defaultArgs,
+      );
 
-    return {
-      exchangeData,
-      needWrapNative: this.needWrapNative,
-      sendEthButSupportsInsertFromAmount: true,
-      dexFuncHasRecipient: false,
-      targetExchange: exchange,
-      returnAmountPos: undefined,
-    };
+      return {
+        exchangeData,
+        needWrapNative: this.needWrapNative,
+        sendEthButSupportsInsertFromAmount: true,
+        dexFuncHasRecipient: false,
+        targetExchange: exchange,
+        returnAmountPos: undefined,
+      };
+    } else {
+      const pathLength = 11;
+      const swapParamsLength = 5;
+      const poolsLength = 5;
+
+      const path = data.path
+        .map((item, index) =>
+          index === 0
+            ? [item.tokenIn, item.exchange, item.tokenOut] // we need tokenIn only for the first item because there is no prev pool
+            : [item.exchange, item.tokenOut],
+        )
+        .flat();
+
+      while (path.length < pathLength) {
+        path.push(NULL_ADDRESS);
+      }
+
+      const swapParams = data.path.map(item => [
+        item.i,
+        item.j,
+        item.underlyingSwap
+          ? CurveRouterSwapType.exchange_underlying
+          : CurveRouterSwapType.exchange,
+        CurveRouterPoolType.stable,
+        item.n_coins,
+      ]);
+
+      while (swapParams.length < swapParamsLength) {
+        swapParams.push([0, 0, 0, 0, 0]);
+      }
+
+      const pools = [];
+
+      while (pools.length < poolsLength) {
+        pools.push(NULL_ADDRESS);
+      }
+
```

Comment:
@aburkut could you clarify this part? not clear why we need to apply such logic to build transaction

Reply:
@KanievskyiDanylo Posted a comment with a link to Curve NG Router exchange function's params description